### DISCUSSION TITLE
To apply the time to the transaction date?

### DISCUSSION THREAD

**Discussion by DevAnyKR** [moneymanagerex/moneymanagerex#5873] [Q&A] [answered]
Can I use only custom fields?

It does not appear in Transaction Grid.

Because there are several transactions in a day, and the order over time is important. Same for sorting.

Additionally, why doesn't the DB use the DATETIME format?
It is currently in the format yyyy-MM-dd.
LASTUPDATE, DELETE Dates are in dateTtime format. 😮

**Comment by tactilis** [comment]
At present in MMEX, transactions have only dates not date_times.

> Can I use only custom fields?

Using a custom Time field would be a way of recording the transaction time.
Alternatively, you could record the time in the Notes field.

> It does not appear in Transaction Grid.

For a custom field to be displayed in the transaction grid, you need to assign the `Panel's column`. If you look at this comment you can see how it is done: https://github.com/moneymanagerex/moneymanagerex/issues/5800#issuecomment-1500339634

<br>

> Additionally, why doesn't the DB use the DATETIME format?
> It is currently in the format yyyy-MM-dd.
> LASTUPDATE, DELETE Dates are in dateTtime format.

Because of MMEX's development history...
Storing just the date of a transaction was an implementation decision made _many_ years ago at the very start of the MMEX project.
The ability to record Last Update and Deletion information is a very recent addition to MMEX. It made sense to make the these fields date_time.

Knowing just the date of a transaction is clearly suitable for the majority of people who use MMEX because there are have only been a couple of questions about storing transaction time over the years.

All the bank data I have ever imported into MMEX has contained only the date of transactions. If there was great demand for date_time then banks would include this as part of their statements and downloaded transaction data.

To help us understand better, could you explain the specific need(s) you have for recording a transaction time please?

**Reply by DevAnyKR** [reply]
Now I know what "UDFC" means. Thanks.
There is a bug where items such as "Panel's column" are initialized (deleted) when I enter Korean in Tooltip.
When Custom Field is applied, the Column position is initialized every time the screen changes.

Financial transaction statements such as banks and cards in Korea include time. Because time is very important. Also, when sorting, it is possible to sort by time, not by scrambling, so it is easy to understand the reason for deposit/withdrawalignment.
I can tell whether the deposit comes first or the expenditure comes first. I can also see what kind of deposit made first, even if I have multiple deposits (expenditure as well)
Transaction records are not only about the fact that it happened, but also about the timing.
This applies when multiple transactions are made in a day.  

If you deny this, the old trading date seems sufficient to record only the year, not the year/month/day. Example: A few years ago, there was a deal 'ABC'.

(Since it is a translated sentence, there may be a misunderstanding)

**Reply by DevAnyKR** [reply]
Of course, I don't need to set the time required when creating a transaction that doesn't matter.
It's either the current time or 00:00:00.

There seems to be no problem migrating the fields in the DB from text to datetime. It's just that the time changes to 00:00:00.

Is this information in the development plan?

**Reply by tactilis** [reply]
> Now I know what "UDFC" means.

I'd agree that it's a not a user-friendly name! It probably stands for `User Defined Field Column`.

Custom fields are a very old feature from when the UI of MMEX was perhaps getting less attention than it is now. We could improve the presentation to the user to make it more obvious.

**Reply by tactilis** [reply]
> Financial transaction statements such as banks and cards in Korea include time. Because time is very important. Also, when sorting, it is possible to sort by time

Understood. It's not common for banks in the UK at least, but I can see that it makes sense to include time.


> There seems to be no problem migrating the fields in the DB from text to datetime. It's just that the time changes to 00:00:00.

That is the smallest part of the change.

There would need to be:
- A change to the database schema
- Changes to the transaction dialogs
- Changes to the filtering dialogs
- Changes to the transaction list display
- Changes to the reports
- Changes to import and export

So actually quite a lot of work.

**Comment by DevAnyKR** [comment]
Depending on the difference between 1 second, the purpose of the transaction is different.
I don't know if I gave you the money first, or if I borrowed the money first and then paid it back later.
TIME records are required. Of course, the transaction's DATETIME default value is yyyy-MM-dd 00:00:00.

It needs to be sorted, but custom fields don't help. And randomly generated fields are at very high risk of being deleted from the database.

**Comment by tactilis** [comment]
@DevelopmentAnything
> Depending on the difference between 1 second, the purpose of the transaction is different.
> I don't know if I gave you the money first, or if I borrowed the money first and then paid it back later.

You can:
1. Enter the transactions into MMEX in the correct order (eg. the deposit when you borrowed the money _and then_ the withdrawal when you pay it back). When you sort by Date, the sorting of the columns is _**always**_ Primary sort: by Date / Secondary sort: by ID.  Thus the transactions on a particular date will _**always**_ be displayed in the order you entered them. This works for CSV import too if the transactions in the data file are ordered from oldest to newest. So the order in MMEX will always be correct, even if you don't have the actual time.

and/or

2. Categorize the transactions to clearly show the purpose. For example, you could do this:

    a) Deposit transaction: `Payee A`, `Borrowing:Receipt of loan`
    b) Withdrawal transaction: `Payee A`, `Borrowing:Repayment of loan`

    c) Withdrawal transaction: `Payee B`, `Lending:Payment of loan`
    d) Deposit transaction:  `Payee B`, `Lending:Receipt of repayment`

<br>

> And randomly generated fields are at very high risk of being deleted from the database.

I'm not sure I understand what you mean by this.

<br>

> It needs to be sorted

It doesn't _need_ to be sorted because it is a fact that most people seem to cope with MMEX using just Date (in fact Date + ID) without any problems. And, as I mentioned, most banks just provide a Date column in their online statements and CSV downloads.

That doesn't mean that it shouldn't be sorted though.

If Korean banks provide date & time, could you please upload a _few_ rows from a couple of real .CSV files please. Include any header rows plus the data rows. It doesn't matter if the language is Korean.  Note that you may need to zip the file for github to allow it to be uploaded.

<br>

As noted previously, there is a lot to change to accommodate time in addition to date: https://github.com/moneymanagerex/moneymanagerex/discussions/5873#discussioncomment-5761340

**Reply by DevAnyKR** [reply]
Written in Korean.

MMEX는 한 개의 계정만 관리하는 프로그램이 아닙니다. 여러 유형의 계정에 여러 거래를 등록할 때, 순서대로 등록할 수 없는 상황이 발생합니다.
오래된 거래 데이터를 수정하거나 누락된 데이터를 입력해야 하는 경우가 있습니다.
오래된 거래를 검색하는 경우도 자주 있습니다.

a. 그럴때마다 거래 시각을 확인하려면 원본 csv 파일을 열어야 합니까?
b. 그럴때마다 mmex 데이터를 초기화 하고, csv 파일을 수정해서 다시 import 해야 합니까?
c. 그럴때마다 mmex에서 순서가 잘못된 거래를 삭제하고 처음부터 다시 등록해야 합니까?

그것은 프로그램을 사용하는 이유가 아닙니다. 종이에 기록을 하고 지우개로 지우는 것과 같습니다.
차라리 ID를 직접 수정할 수 있게 하는 편이 낫겠습니다.

> Categorize the transactions to clearly show the purpose. For example, you could do this:

MMEX 거래 데이터는 7천개가 넘습니다. 그 중 수정이 필요한 데이터를 찾고 수정하는 것은 상상하고 싶지 않습니다.

> I'm not sure I understand what you mean by this.

사용자 정의 필드는 사용자가 언제든지 변경하거나 삭제할 수 있습니다. 하지만 거래 날짜 필드는 삭제할 수 없습니다. 사용자 정의 필드는 데이터 안정성에 취약합니다.

> I mentioned, most banks just provide a Date column in their online statements and CSV downloads.

정렬은 하나의 계정만 포함하는 것이 아닙니다. 모든 거래에서는 모든 계정을 정렬 하기 때문에 은행의 시간 기록 요구는 의미가 없습니다. 왜냐하면 MMEX는 은행 뿐만 아니라 신용 카드, 페이, 포인트와 같은 여러 종류의 계정 유형을 함께 기록하는 프로그램이기 때문입니다.

은행 거래 내역은 CSV 대신 이미지로 대체하겠습니다. 저는 스마트폰 앱으로 거래 내역을 수신합니다.
기억하세요! MMEX에는 은행 거래 내역만 등록하는 프로그램이 아닙니다!
![Screenshot_20230718-001257_KB](https://github.com/moneymanagerex/moneymanagerex/assets/110871727/3dd93d03-4290-46df-be8b-3ad055f71707)

> It doesn't need to be sorted because it is a fact that most people seem to cope with MMEX using just Date (in fact Date + ID) without any problems.

대부분의 사람들이 문제 없이 사용한다는 것은 동의할 수 없습니다. 실제 MMEX를 사용하는 사람 중, 몇 %가 피드백을 합니까?
저는 많은 프로그램을 사용하면서 문제점이나 요구사항을 피드백한 경우가 거의 없습니다.
 
6. 지금 당장 V1.6에 기능을 추가해 달라는 것이 아닙니다. V1.7이 될 수도 있습니다. 하지만 거래 이력에 시간은 중요한 요소입니다.
계획이 없다면 한국어는 삭제해 주셔도 됩니다. 한국에는 이미 여러 종류의 가계부(자금관리) 프로그램이 있으며 모두다 거래 시각 기록이 가능합니다.

**Reply by tactilis** [reply, marked as answer]
> (Auto translated from Korean)
> What percentage of people who actually use MMEX give feedback?

Very few compared to the number of instances of MMEX installed. Only those with problems raise issues. As I said, the majority seem to cope OK.

<br>

> (Auto translated from Korean)
> I've used a lot of programs and rarely have feedback on issues or requirements.

_Exactly._ That's the point I'm making.

<br>

> (Auto translated from Korean)
> If you don't have a plan, you can delete the Korean language.
> There are already several types of household finance (money management) programs in Korea, and all of them can record transaction times.

Maybe those other apps would be more suited to your needs?

<br>

As I said earlier regarding the addition of a time field:
> _[It] doesn't mean that it shouldn't be sorted though._

I understand your arguments, however, there is no 'plan' at the moment.

It's a large amount of work which impacts multiple areas of MMEX. Whether it gets done (or not) will be dependant on whether a volunteer developer is personally interested enough to want to devote the time and expend the effort required to implement this.

**Reply by DevAnyKR** [reply]
thank you
I understand that there are no plans for versions after V2.0.
I don't know C++ so I can't help. I apologize on behalf of anyone who needs this feature.
It is said that there were problems when MMEX was developed with C# in the early days, but it seems that problems such as slowness can be sufficiently improved if the latest C# version is used. This is a separate project worth starting over.
